strawberryfields/tdm/tdmprogram.py

```diff
@@ -360,6 +396,24 @@ def apply_op(self, cmd, q, t):
 
         self.append(cmd.op.__class__(*params), get_modes(cmd, q))
 
+    def assert_number_of_modes(self, device):
+        """Check that the number of modes in the program is valid for the given device."""
+        if self.timebins > device.modes["temporal"]["max"]:
+            raise CircuitError(
+                f"This program contains {self.timebins} temporal modes, but the device '{device.target}' "
+                f"only supports up to {device.modes['temporal']['max']} modes."
+            )
```

Comment:
I think this suggestion is worth doing

Reply:
I'm not sure if that's actually a better solution, looking closer at it. I think it's fine having three checks, and separate error raises, since they differ in either having to not be `!=` or `>` the temporal/concurrent/spatial modes value in the device spec.

Reply:
The original suggestion aimed at only factoring out the string for the message, but keeping the three checks separated.

Inside of the string, the only difference seems to be the type of the mode (e.g., `This program contains {self.timebins} temporal modes`, `This program contains {self.concurr_modes} concurrent modes`, etc.). Here we could go with `This program contains {num_modes} {mode_type} modes`. It just aimed at transitioning to have a single string that we'd have to maintain. Although it involves moving from using `f` strings to using `format`, think that could be fine.

Reply:
I see your point (and I almost agree with it). But there's the slight extra difference that one of the messages has the additional `up to` in it. If you're fine with it, I might just keep this as it is. I think it works fine, and is the simplest solution without having to re-word it in a slightly different way.